kong/pdk/response.lua

```diff
@@ -566,6 +585,15 @@ local function new(pdk, major_version)
       validate_headers(headers)
     end
 
+    if status >= 500 then
```

Comment:
Agreed with what @bungle said. It also seems like by trying to mimic the old `kong.tools.responses` module, we are introducing some undocumented and arbitrary behavior (i.e. that a table with a `message` key will log the error).

I'd suggest that we explicitly log such errors ourselves (yes, from the caller's side, which is the plugin), and later introduce a higher-level function which can provide more flexibility; something like this:
```lua
kong.response.exit_log_err(status, body, headers, log_msg, log_lvl)
```

Which would be called like:
```lua
local ok, err = do_thing()
if not ok then
  local body = {
    err_code = 123,
    message = "This is a custom Bad Gateway response payload"
  }

  return kong.response.exit_log_err(kong.response.BAD_GATEWAY, body, nil,
                                    err, kong.log.CRIT)
end
```

Keep in mind this is an early proposal, but note how providing a higher-level API gives us:

* the ability to send a response without logging, if undesired
* more explicitness as to what the function is doing (although since we just added some arguments, we could drop this benefit, and provide this in `kong.response.exit` when more arguments are used, it wouldn't be too bad)
* flexibility as to what the user-facing message is (it can be overridden compared to the default `BAD GATEWAY` message which would just be "Bad Gateway")
* distinction between the user-facing and admin-facing error messages are
* ability to customize the admin-facing log level (maybe some errors from some plugins e.g. OAuth2/OIDC warrant an `ALERT` log level, so as to wake-up somebody on call? Maybe some other errors could just be warnings too).

Reply:
Points taken, I will revert this in this PR and add the explicit log calls. We can get around to design/implement the enhanced exit+log in a future PR.